app/models/product_import/entry_validator.rb

```diff
@@ -189,6 +196,43 @@ def unit_fields_validation(entry)
                              error: I18n.t('admin.product_import.model.conditional_blank'))
     end
 
+    def is_numeric(value)
+      return true unless Float(value, exception: false).nil?
+    end
+
+    def price_validation(entry)
+      return if is_numeric(entry.price)
+
+      if empty_or_placeholder_value(entry.price)
+        mark_as_invalid(entry, attribute: 'price',
+                               error: I18n.t('admin.product_import.model.blank'))
+      else
+        mark_as_invalid(entry, attribute: 'price',
+                               error: I18n.t('admin.product_import.model.incorrect_value'))
+      end
+    end
+
+    def on_hand_on_demand_validation(entry)
+      on_hand_present_numeric = !empty_or_placeholder_value(entry.on_hand) &&
+                                is_numeric(entry.on_hand)
+      on_hand_value = entry.on_hand&.to_i
+      on_demand_present_numeric = !empty_or_placeholder_value(entry.on_demand) &&
+                                  is_numeric(entry.on_demand)
+      on_demand_value = entry.on_demand&.to_i
+
+      return if (on_hand_present_numeric && on_hand_value >= 0) ||
+                (on_demand_present_numeric && on_demand_value == 1)
+
+      mark_as_invalid(entry, attribute: 'on_hand',
+                             error: I18n.t('admin.product_import.model.incorrect_value'))
+      mark_as_invalid(entry, attribute: 'on_demand',
+                             error: I18n.t('admin.product_import.model.incorrect_value'))
+    end
```

Comment:
I don't see any advantage to validating these two fields in the same method, so could it be split into two?

Reply:
I initially tried having separate validation methods for them. 
But as we had to read both of the values in both the methods(for the relation between on_hand and on_validation), it seemed repetitive.

Eg: 

```
def on_hand_validation(entry)
  on_hand_present_numeric = !empty_or_placeholder_value(entry.on_hand) &&
                            is_numeric(entry.on_hand)
  on_hand_value = entry.on_hand&.to_i

  return if on_hand_present_numeric && on_hand_value >= 0 ||
            (!empty_or_placeholder_value(entry.on_demand) &&
             entry.on_demand&.to_i == 1)

  mark_as_invalid(entry, attribute: 'on_hand',
                         error: I18n.t('admin.product_import.model.incorrect_value'))
end

def on_demand_validation(entry)
  on_demand_present_numeric = !empty_or_placeholder_value(entry.on_demand) &&
                              is_numeric(entry.on_demand)
  on_demand_value = entry.on_demand&.to_i

  return if on_demand_present_numeric && on_demand_value == 1 ||
            (!empty_or_placeholder_value(entry.on_hand) &&
             entry.on_hand&.to_i >= 0)

  mark_as_invalid(entry, attribute: 'on_demand',
                         error: I18n.t('admin.product_import.model.incorrect_value'))
end

```

Having said that, i've reduced the validation to the relation between the fields,i.e, if at least one of them has a valid value, other one is ignored regardless of its value.

Eg: **Ignoring cases like entry 2 and 3**
| Entry     | on_hand | on_demand | Output | Ideal output |
|-----------|---------|-----------|-----------------|--|
| Entry 1   | 10      | 1         | Valid           | Valid|
| Entry 2   | -5      | 1         | Valid           | Invalid on_hand|
| Entry 3   | foo     | 1         | Valid           | Invalid on_hand|
| Entry 4   | 10      | 0         | Valid           | Valid|
| Entry 5   | -5      | 0         | Invalid           |Invalid|
| Entry 6   | foo     | 0         | Invalid           |Invalid|
| Entry 7   |          | 1         | Valid           |Valid|
| Entry 8   | -10     |           | Invalid           |Invalid|
| Entry 9   |          |           | Invalid         |Invalid|
| Entry 10  | 5     | 1         | Valid         |Valid|
| Entry 11  | 5     |           | Valid         |Valid|
| Entry 12  | -5      |           | Invalid         |Invalid|

Reply:
Also while adding tests, I learned that they have to be mutually exclusively enabled for importing type as "Inventories".
Added more test covering this case as well.

Reply:
Ah, thanks for responding, I didn't pick up on the relationship between the two.